src/core/net/dns_client.cpp

```diff
@@ -815,20 +862,76 @@ void Client::SendQuery(Query &aQuery, QueryInfo &aInfo, bool aUpdateTimer)
         SuccessOrExit(error = message->Append(Question(kQuestionRecordTypes[aInfo.mQueryType][num])));
     }
 
-    messageInfo.SetPeerAddr(aInfo.mConfig.GetServerSockAddr().GetAddress());
-    messageInfo.SetPeerPort(aInfo.mConfig.GetServerSockAddr().GetPort());
+    length = message->GetLength() - message->GetOffset();
+#if OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE
 
-    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
+    // Send udp query in case of message being greater than kUdpQueryMaxSize or as last resort.
+    if (aInfo.mConfig.GetTransportProto() == QueryConfig::kDnsTcpTransport)
+    {
+        mSendLink.mNext = nullptr;
+        mSendLink.mData = mSendBufferBytes;
 
-exit:
-    FreeMessageOnError(message, error);
+        // Prepending the DNS query with length of the packet according to RFC1035.
+        WriteUint16(length, mSendBufferBytes);
+        message->Read(message->GetOffset(), (mSendBufferBytes + sizeof(uint16_t)), length);
+        mSendLink.mLength = length + sizeof(uint16_t);
 
-    UpdateQuery(aQuery, aInfo);
+        mTcpTimer.Stop();
+        switch (mTcpState)
+        {
+        case kTcpUninitialized:
+            mCurrentTcpQuery = &aQuery;
+            SuccessOrExit(InitTcpSocket());
+            SuccessOrExit(mEndpoint.Connect(AsCoreType(&aInfo.mConfig.mServerSockAddr), OT_TCP_CONNECT_NO_FAST_OPEN));
+            break;
+        case kTcpInitialized:
+            if (memcmp(&mEndpoint.GetPeerAddress(), &aInfo.mConfig.mServerSockAddr, sizeof(otSockAddr)) != 0)
+            {
+                error = kErrorFailed;
+                ExitNow();
+            }
+            mCurrentTcpQuery = &aQuery;
+            mTcpState        = kTcpSendingInProgress;
+            SuccessOrExit(error = mEndpoint.SendByReference(mSendLink, /* aFlags */ 0));
+            break;
+        case kTcpSendingInProgress:
+            if (memcmp(&mEndpoint.GetPeerAddress(), &aInfo.mConfig.mServerSockAddr, sizeof(otSockAddr)) != 0)
+            {
+                error = kErrorFailed;
+                ExitNow();
+            }
+            // Decrease number of attempts on TCP query since this send will be handled in send done.
+            aInfo.mTransmissionCount--;
+            break;
+        default:
+            OT_ASSERT(false); // Should never get here.
+            break;
+        }
+        aInfo.mRetransmissionTime = TimerMilli::GetNow() + aInfo.mConfig.GetResponseTimeout();
+        mTcpTimer.FireAt(aInfo.mRetransmissionTime);
+        message->Free();
+        message = nullptr;
```

Comment:
@abtink I have introduced a state machine for handling TCP connection. So now it's possible to send multiple queries without altering the buffer.
So a few words of explanation of how it works:
- `kTcpUninitialized` means that the connection is not operational
- `kTcpConnecting` means that connection establishment is in progress next query will be sent in `HandleTcpEstablishedCallback`
- `kTcpInitialized`  - means that the connection was established but is idle.
- `kTcpSendingInProgress` - means the query is in progress and we are queuing the next to the same server to be sent in `HandleTcpSendDoneCallback`

Reply:
Also, the `SendQuery` method now returns a value to fail fast in case we would try to send a TCP query to a different server.

Reply:
Both sounds good to me. 👍